internal/controller/config.go

```diff
@@ -0,0 +1,123 @@
+package controller // import "go.opentelemetry.io/ebpf-profiler/internal/controller"
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"runtime"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/ebpf-profiler/tracer"
+)
+
+type Config struct {
+	BpfVerifierLogLevel    uint
+	CollAgentAddr          string
+	Copyright              bool
+	DisableTLS             bool
+	MapScaleFactor         uint
+	MonitorInterval        time.Duration
+	ClockSyncInterval      time.Duration
+	NoKernelVersionCheck   bool
+	PprofAddr              string
+	ProbabilisticInterval  time.Duration
+	ProbabilisticThreshold uint
+	ReporterInterval       time.Duration
+	SamplesPerSecond       int
+	SendErrorFrames        bool
+	Tracers                string
+	VerboseMode            bool
+	Version                bool
+
+	Fs *flag.FlagSet
+}
+
+const (
+	exitParseError int = 2
+
+	// 1TB of executable address space
+	MaxArgMapScaleFactor = 8
+)
+
+// Dump visits all flag sets, and dumps them all to debug
+// Used for verbose mode logging.
+func (cfg *Config) Dump() {
+	log.Debug("Config:")
+	cfg.Fs.VisitAll(func(f *flag.Flag) {
+		log.Debug(fmt.Sprintf("%s: %v", f.Name, f.Value))
+	})
+}
+
+// Validate runs validations on the provided configuration, and returns errors
+// if invalid values were provided.
+func (cfg *Config) Validate() error {
+	if cfg.SamplesPerSecond < 1 {
+		return ErrorWithExitCode{
+			error: fmt.Errorf("invalid sampling frequency: %d", cfg.SamplesPerSecond),
+			code:  exitParseError,
+		}
```

Comment:
There is a lot of code repetition here. The old code uses a helper function `parseError()`, which can be easily adopted.

Reply:
Returning an exit code is quite weird, and not really a Go convention. Hence this change.

Reply:
I could make things slightly easier to read with a `parseError()` method that returns an `ErrorWithExitCode`?

Reply:
See 51713d3c07e3a2dab45301a12a4ab0546fb0fad7

Reply:
You can also move `fmt.Errorf` into `parseError` and only pass the format+args.
Optionally, you can return `ErrorWithExitCode` from Validate and not have to type cast check in main.go.

Reply:
Sure, though it's not "parseError" anymore. But that removes a bit of repetition.
However, I don't think returning `ErrorWithExitCode` is great. `Validate()` returning a standard type makes it more generic. The fact that is can provide more context with an exit code is an implementation detail that doesn't need to leak to the method signature.

Reply:
> Validate() returning a standard type makes it more generic. The fact that is can provide more context with an exit code is an implementation detail that doesn't need to leak to the method signature.

That are fair points. Though the downside of this is that the caller has to take care for runtime casting and type checking with something "ugly" like
```
	if err, ok := err.(controller.ErrorWithExitCode); ok {
		return exitCode(err.Code())
```
But this is not meant as a blocker for the PR :)